### DISCUSSION TITLE
Forcing 'lfs migrate' on 'remote' branches

### DISCUSSION THREAD

**Discussion by chriswarkentin** [git-lfs/git-lfs#5299] [General]
I'm migrating a huge 20 year old SVN repository to Git via git-svn which is going well except for one problem. This might sound like a git-svn problem but I hope you'll hear me out.

Since we're going to be hosting on GitHub I have a cleanup process that strips a bunch of large binaries from the distant past with BFG and then does a 'git lfs migrate' to put everything else into LFS. This works exceptionally well and I'm left with a very clean repository. If we were doing a one-shot migration it would be done but we need to maintain the link to SVN for a period of time during the migration.

In order for the above cleanup to work I need to check out the branches locally before migration because git-lfs won't migrate branches that it thinks are remote. Once that happens the branches are 'diverged' but I don't know how to get the new, migrated branches back to where git-svn wants them.

The reason I put 'remote' in quotes is that they LOOK like real remotes:

> $ git branch -a | grep feature_tls_logs
>   feature_tls_logs
>   remotes/origin/feature_tls_logs

But they aren't, exactly. That is to say, I can't do this:

> $ git push -f origin feature_tls_logs
> fatal: 'origin' does not appear to be a git repository
> fatal: Could not read from remote repository.
> 
> Please make sure you have the correct access rights
> and the repository exists.

I believe in theory it should be possible so perhaps that's a more generic question for Stack Overflow.

Is there a way to force git-lfs migrate to work on these weird git-svn branches? BFG seems to do it with no issues and then I can keep pulling from svn (after some minor tweakery). The other would be if I could figure out what the heck these branches are and how to recreate them but that's a git-svn/stack-overflow issue, unless some expert here has an idea.

Thanks in advance for reading this slightly long winded question. I'd be willing to make a local 'hack' to git-lfs to force it to work. The alternative is going through a massive pile of perl that is git-svn to figure out what it's doing.

**Comment by bk2204** [comment]
Hey,

I believe the `--everything` option for `git lfs migrate import` should operate on remote branches as well as local branches.  However, I can't speak for how it works with git-svn, but you're welcome to try.  My guess is that trying to do so won't work because SVN doesn't allow rewriting older commits, but you can certainly try.

**Reply by chriswarkentin** [reply]
Hi, thanks for replying. It doesn't seem to migrate the remote branches at all. The documentation says that --everything will not hit anything it thinks is a remote branch. As I said, it's not REALLY a remote branch but I'm not sure what it is. All I know is that BFG definitely processes them so I'm not sure why git lfs migrate doesn't.

> $ git diff feature_tls_logs origin/feature_tls_logs
> diff --git a/.gitattributes b/.gitattributes
> deleted file mode 100644
> index c65eeaa40fc..00000000000
> --- a/.gitattributes
> +++ /dev/null
> @@ -1,250 +0,0 @@
> -*.a filter=lfs diff=lfs merge=lfs -text
> -*.at filter=lfs diff=lfs merge=lfs -text
> -*.bin filter=lfs diff=lfs merge=lfs -text
> -*.bmp filter=lfs diff=lfs merge=lfs -text
> -*.bz2 filter=lfs diff=lfs merge=lfs -text
> -*.chm filter=lfs diff=lfs merge=lfs -text

**Reply by chriswarkentin** [reply]
Hmm, looking through the code now and this looks promising. command_migrate.go:235
```
		if !bare {
			// Otherwise, if neither --include-ref=<ref> or
			// --exclude-ref=<ref> were given, include no additional
			// references, and exclude all remote references that
			// are remote branches or remote tags.
			remoteRefs, err := getRemoteRefs(l)
			if err != nil {
				return nil, nil, err
			}

			for remote, refs := range remoteRefs {
				for _, ref := range refs {
					exclude = append(exclude,
						formatRefName(ref, remote))
				}
			}
		}
```

**Reply by chriswarkentin** [reply]
Oh...I wonder if I were to run this in the .git folder what would happen?

**Reply by chriswarkentin** [reply]
FWIW my command line is:

> git lfs migrate import --everything --yes --verbose --skip-fetch --include="$patterns"

where 'patterns' is rather a lot of patterns...

**Reply by chriswarkentin** [reply]
Looking at the larger context, it really looks like you're right and it SHOULD be migrating everything. I don't know why it isn't. I might need to do some tests.

**Reply by chriswarkentin** [reply]
In case anyone is still reading I found the issue. The migration DOES in fact do all the branches but fails to update the references. In the general sense this is PROBABLY the right thing to do since you're expected to force-push to the remote later. In my case since there isn't actually a real remote, I just commented out this section and all was well.
```
diff --git a/git/githistory/rewriter.go b/git/githistory/rewriter.go
index 3ab095cc..4f7eeb6d 100644
--- a/git/githistory/rewriter.go
+++ b/git/githistory/rewriter.go
@@ -550,10 +550,12 @@ func (r *Rewriter) refsToMigrate() ([]*git.Ref, error) {

        var local []*git.Ref
        for _, ref := range refs {
+               /*
                        if ref.Type == git.RefTypeRemoteBranch {
                               continue
                        }
+               */
                local = append(local, ref)
        }
```

**Reply by bk2204** [reply]
Ah, yes, that makes sense.  Since we use the remote tracking branches as an optimization to not push objects the remote already has, that would mean we wouldn't push anything, since the remote side would be considered already up to date.

I'm glad you got it figured out.

**Reply by chriswarkentin** [reply]
Yes, thanks for the help. Technically a flaw in BFG in my opinion although it actually worked out for me in this case.
